### ISSUE TITLE
Will Keras output non-binary (continuous) predictions for a classification problem?

### ISSUE THREAD

**Issue by mpt5** [fchollet/keras#3173]
My current model is compiled as

    model.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=['accuracy'])
and I get predictions like
```
[[ 1.  0.  0.  0.]]
[[ 0.  1.  0.  0.]]
```
Instead I'd like my class predictions to show a percentage
    
     [[ 0.8  0.1  0.1  0.4]]

Elsewhere I was advised to use `loss='mse'`, but this was discouraged by another user.
This issue does not seem addressed in the docs (I'd be glad to add to docs any insight you can give).

Any advice on what is best practice using Keras?

**Comment by mbollmann** [comment]
If you're using `predict()` to generate your predictions, you should already get probabilities (provided your last layer is a softmax activation), nothing to do with the loss function. Are you sure you're using `predict()`, not `predict_classes()`?

**Comment by mpt5** [comment]
@mbollmann thanks, you're correct -- it looks like despite using `predict` my output was ones and zeroes even in cases that were incorrectly classified (in these I expected a fractional output due to uncertainty). As you pointed out, I am using `softmax`.

**Comment by alyato** [comment]
@mbollmann  I check the predict_classes.
![default](https://cloud.githubusercontent.com/assets/6815071/17251142/b911582c-5564-11e6-891f-40b6f43d0aeb.png)
I don't use the function of `predict_classes` and use the `predict` function
Then i don't know what does the `predict_classes` return.
Does it be the format,like this.

> [[ 1.  0.  0.  0.]]
[[ 0.  1.  0.  0.]]

I also don't know when use the function of `predict_classes`?
Thanks.

**Comment by mbollmann** [comment]
@alyato `predict_classes` treats the output of the model as class probabilities, and instead of returning the probabilities, it returns the index with the highest probability.

You can just try both functions and look at their output to see the difference.

**Comment by root-master** [comment]
@mptorr I have the same problem. `model.predict()` returns predictions (zeros and ones) and not the probability (although every document on Keras states the opposite).

**Comment by hskang9** [comment]
How about using sigmoid as the activation function for the neural network classifying at the very end. Since sigmoid is differentiable, it returns value 1 or other values(0.39, 9.99, etc).

**Comment by mandarup** [comment]
@root-master have you resolved this? I am also facing the same issue `predict` and `predict_classes` give {0,1}

**Comment by AMFIRNAS** [comment]
predict_proba() is not there for functional api, I think. So how to get the predicted real values if it a regression problem in lstm?

**Comment by ShunyuanZ** [comment]
In my case, model.predict() and model.predict_proba() both give probabilities. model.predict_classes() give binary responses

**Comment by xiaoleitw** [comment]
try normalization on your test set and you may get the probabilities.

**Comment by Saharati90** [comment]
make sure your target is categorized:
>>labels = [0,1,1,1,0]
>>labels = to_categorical(labels,num_classes = None)
labels = [[1,0],[0,1],[0,1],[0,1],[1,0]]

and then:
model.add(Denese(2,activation= 'softmax'))
ans then:
model.fit(X_train,label,...)
model.predict(X_test)

**Comment by garwahl** [comment]
@root-master , was there any resolution to this issue?

**Comment by coderefaj** [comment]
Any Solution To this ?

**Comment by coderefaj** [comment]
You can do it with by migrating to previous layer in model . before final layer . 
It will show the prob ,

**Comment by shamim-pku** [comment]
Just change the data types to float32 and normalize it the following way then the results will give the probabilities
Data = X_train.astype('float32')
Data /=255
...
model.predict_proba(Data, verbose=1)

**Comment by thefatty0418** [comment]
> Just change the data types to float32 and normalize it the following way then the results will give the probabilities
> Data = X_train.astype('float32')
> Data /=255
> ...
> model.predict_proba(Data, verbose=1)

thanks, it really help and solve the problem!
